starport/services/network/join.go

```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return nil, err
+	}
+	if exist {
+		return nil, errors.New("validator already exist: " + valAddress)
+	}
+
+	return launchtypes.NewMsgRequestAddValidator(
+		valAddress,
+		launchID,
+		gentx,
+		consPubKey,
+		selfDelegation,
+		peer,
+	), nil
+}
+
+// CreateAccountRequestMsg creates an add AddAccount request message
+func (b *Builder) CreateAccountRequestMsg(
+	ctx context.Context,
+	chainHome string,
+	launchID uint64,
+	amount sdk.Coin,
+) (msg sdk.Msg, err error) {
+	addr := b.account.Address(SPNAddressPrefix)
+	b.ev.Send(events.New(events.StatusOngoing, "Verifying account already exists "+addr))
+
+	shouldCreateAcc := false
+	if !amount.IsZero() {
+		exist, err := CheckGenesisAddress(chainHome, addr)
+		if err != nil {
+			return msg, err
+		}
+		if !exist {
+			exist, err = b.CheckRequestAccount(ctx, launchID, addr)
+			if err != nil {
+				return msg, err
+			}
+		}
+		shouldCreateAcc = !exist
+	}
+	if shouldCreateAcc {
+		b.ev.Send(events.New(events.StatusDone, "Account message created"))
+		msg = launchtypes.NewMsgRequestAddAccount(
+			addr,
+			launchID,
+			sdk.NewCoins(amount),
+		)
+	} else {
+		b.ev.Send(events.New(events.StatusDone, "Account message not created"))
+	}
+	return msg, err
+
+}
+
+// GetAccountAddress return an account address for the blockchain by name
+func (b *Blockchain) GetAccountAddress(ctx context.Context, accountName string) (string, error) {
+	if !b.isInitialized {
+		return "", errors.New("the blockchain must be initialized to show an account")
+	}
+
+	chainCmd, err := b.chain.Commands(ctx)
+	if err != nil {
+		return "", err
+	}
+	acc, err := chainCmd.ShowAccount(ctx, accountName)
+	if err != nil {
+		return "", err
+	}
+	return acc.Address, nil
+}
+
+// CheckRequestAccount check if the add account request already exist
+func (b *Builder) CheckRequestAccount(ctx context.Context, launchID uint64, addr string) (bool, error) {
+	requests, err := b.fetchRequests(ctx, launchID)
+	if err != nil {
+		return false, err
+	}
+	for _, request := range requests {
+		genesisAcc := request.Content.GetGenesisAccount()
+		if genesisAcc == nil {
+			continue
+		}
+		if genesisAcc.Address == addr {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// checkRequestValidator check if the add validator request already exist
+func (b *Builder) checkRequestValidator(ctx context.Context, launchID uint64, addr string) (bool, uint64, error) {
+	requests, err := b.fetchRequests(ctx, launchID)
+	if err != nil {
+		return false, 0, err
+	}
+	for _, request := range requests {
+		genesisVal := request.Content.GetGenesisValidator()
+		if genesisVal == nil {
+			continue
+		}
+		if genesisVal.Address == addr {
+			return true, genesisVal.LaunchID, nil
+		}
+	}
+	return false, 0, nil
+}
+
+// fetchRequests fetches the chain requests from SPN by launch id
+func (b *Builder) fetchRequests(ctx context.Context, launchID uint64) ([]launchtypes.Request, error) {
```

Comment:
These are not the requests we should fetch to check account existence but the already approved accounts and validators and therefore querying the requests `GenesisAccountAll`, `GenesisAccountAll` and `GenesisValidatorAll`